Added preliminary PHP support.

kiuby88 opened the pull request:
PHP support is work-in-progress. It allows deploying of PHP web applications from git repositories and tarball sources. 

Below, we can see a  php deployment example using a git repo.

```
name: HelloWorld
services:
- serviceType: brooklyn.entity.webapp.apache.ApacheServer
 name: Apache Server
 location: localhost
 http_port: 80
 app_git_repo_url: https://github.com/kiuby88/phpHelloWorld.git
```

The next plan shows a php deployment example using a tarball resource

```
name: HelloWorld
services:
- serviceType: brooklyn.entity.webapp.apache.ApacheServer
 name: Apache Server
 location: localhost
 http_port: 80
 app_tarball_url : https://github.com/kiuby88/phpHelloWorld.git 
```

Known issues:
- This implementation deploys an Apache Web Server. No application server support yet. 
- Metrics will be provided by Apache mod_status module.
- No HTTPS support yet.

Commit:
Adding php support. Several generic classes can be find  which are
concreted using apache server. The apache specification have to be change
by something like PhpHttpd or similar.
The php application can be specified like a git repo or a tarball
ressource.
A live test is included.

andreaturli commented on software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:
```diff
@@ -740,4 +740,72 @@ protected ScriptHelper newScript(Map<String, ?> flags, String phase) {
     @Override
     public void setup() { }
 
+
```
General comment: I don't think the `AbstractSoftwareProcessSshDriver` needs to install always git. Can you move this code down to `PhpWebAppSshDriver` maybe?

kiuby88 replied:
Yea, but I added this method to `AbstractSoftwareProcessSshDriver` as a general funtionality and if a different entity needed install git, it would not need re-write this code. What do you think?

andreaturli commented on software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:
```diff
@@ -740,4 +740,72 @@ protected ScriptHelper newScript(Map<String, ?> flags, String phase) {
+        return result;
+    }
+
+    private void checkAndInstallGit(){
```
I'd move that logic to PhpWebAppSshDriver.install

kiuby88 replied:
Same here, this method provides how download  (clone) a git repo using a target Folder in a generic way. All entity that need clone a git repo are able to use this method.

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java:
```diff
@@ -0,0 +1,203 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
```
try to avoid wildcards

kiuby88 replied:
I think so, thx.

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcess.java:
```diff
@@ -0,0 +1,62 @@
+    ConfigKey<String> SUGGESTED_VERSION= ConfigKeys.newConfigKeyWithDefault(SoftwareProcess.SUGGESTED_VERSION, "5");
+
+    /**
+     * So, the download method have to define the path.
```
Could you rephrase this comment?

kiuby88 replied:
This comment and the next are re-written.

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java:
```diff
@@ -0,0 +1,203 @@
+
+    private String inferCorrectAppGitName(){
+        String result;
+        if(getConfig(APP_NAME)==null || getConfig(APP_NAME).equals("")){
```
Maybe better `Strings.isEmpty(getConfig(APP_NAME))` ?

kiuby88 replied:
done

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java:
```diff
@@ -0,0 +1,203 @@
+
+    private String  inferCorrectAppTarballName(){
+        String result;
+        if(getConfig(APP_NAME)==null || getConfig(APP_NAME).equals("")){
```
Maybe better `if (Strings.isEmpty(getConfig(APP_NAME)))` ?

kiuby88 replied:
done

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java:
```diff
@@ -0,0 +1,203 @@
+        return result;
+    }
+
+    @Effector(description="Deploys the given artifact, from a source URL, to a given deployment filename/context")
```
move the `@Effector` annotation up to the `PhpWebAppSoftwareProcess` interface

kiuby88 replied:
The annotations are duplicate, so, I am going to remove this copy.

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java:
```diff
@@ -0,0 +1,203 @@
+        setDeployedPhpAppsAttribute(deployedPhpApps);
+    }
+
+    @Effector(description="Deploys the given artifact, from a source URL, to a given deployment filename/context")
```
same here- move to the interface the annotation

kiuby88 replied:
Same above

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSshDriver.java:
```diff
@@ -0,0 +1,193 @@
+    @Override
+    public void install() {
+        //Fixme use the newScript to install php
+//        List<String> commands = ImmutableList.<String>builder()
```
consider to move here the commands to install git

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java:
```diff
@@ -0,0 +1,106 @@
+import brooklyn.event.basic.Sensors;
+import brooklyn.util.flags.SetFromFlag;
+
+@Catalog(name="Apache Web Server", description="ApacheServer:  is a web server application which can be connect" +
```
Similarly to `Tomcat` entity, I think we should name it 

`@Catalog(name="HTTP Server", description="Apache HTTP Server Project is an open-source HTTP server")` 

as described in the official website http://httpd.apache.org/

andreaturli commented on software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java:
```diff
@@ -0,0 +1,106 @@
+@Catalog(name="Apache Web Server", description="ApacheServer:  is a web server application which can be connect" +
+        "with ", iconUrl="classpath:///jboss-logo.png")
+@ImplementedBy(ApacheServerImpl.class)
+public interface ApacheServer extends PhpWebAppSoftwareProcess, PhpWebAppService, HasShortName {
```
I'm not completely sure HttpServer should always extend PhpWebApp*
Unless it is a customized entity for Seaclouds, we should make the PHP support optional

@ahgittin what do you think?

kiuby88 replied:
@andreaturli 
I defined this architecture because I thought that it could be extensible. For example, for deploy php using JBoss only could be necessary implements a JBoss which will extend PhpWebApp*. However, currently I think that it is not the best point of view, Apache should be an independent entity which should be used for PhpWebApp for deploying the artifacts.
So, I think your question is a good idea and I would like know your thinking about this.

Commit:
Minor fix of pull request issues.
Imports.
Some comments are re-written.
Some conditions are refactored.
Duplicate effector notations are removed.
Solve the error: A configKey can not be cast to String
Definition of the two Double Sensors have been modified
Mysql support was added for php5
